### PR TITLE
Option for decimal output

### PR DESCRIPTION
**Pull request by akobler**
Hi

Thanks for the great library! I had the need to encode HTML entities to the decimal format instead of hexadecimal format. Please review the changes. Are additional test cases necessary? As I am not an expert in HTML entities, I can not really judge about corner cases that must be tested.

Thanks for reviewing and (hopefully) merging back.

Andi

### CODE REVIEW

**Comment by akobler** [comment]
The option might be renamed to useDecimalReferences? (to be in line with useNamedReferences)

File: bin/he [hunk 1]
```diff
@@ -64,6 +64,11 @@
 				options.allowUnsafeSymbols = true;
 				return;
 			}
+			if (string == '--decimal-output') {
+				action = 'encode';
+				options.decimalOutput = true;
+				return;
+			}
```
Comment (mathiasbynens) [inline bin/he]: Let’s rename the flag to `--decimal`.

File: man/he.1 [hunk 2]
```diff
@@ -1,4 +1,4 @@
-.Dd July 27, 2013
+.Dd July 27, 2013&#xA9;
```
Comment (mathiasbynens) [inline man/he.1]: ?

File: src/he.js [hunk 3]
```diff
@@ -210,14 +216,15 @@
...
+			.replace(regexBmpWhitelist, escapeSymbol);
 	};
 	// Expose default options (so they can be overridden globally).
 	encode.options = {
 		'allowUnsafeSymbols': false,
 		'encodeEverything': false,
 		'strict': false,
-		'useNamedReferences': false
+		'useNamedReferences': false,
+		'decimalOutput' : false
```
Comment (mathiasbynens) [inline src/he.js]: Let’s rename the setting to just `decimal`.

File: tests/tests.js [hunk 4]
```diff
@@ -6620,6 +6620,11 @@
 			Error,
 			'Parse error: forbidden code point when `allowUnsafeSymbols: true` and `strict: true`'
 		);
+		equal(
+			he.encode('äöüÄÖÜ', { 'decimalOutput': true}),
```
Comment (mathiasbynens) [inline tests/tests.js]: Please use escapes sequences here, matching the other tests: [`'\xE4\xF6\xFC\xC4\xD6\xDC'`](https://mothereff.in/js-escapes#1%C3%A4%C3%B6%C3%BC%C3%84%C3%96%C3%9C)

File: tests/tests.src.js [hunk 5]
```diff
@@ -6620,6 +6620,11 @@
 			Error,
 			'Parse error: forbidden code point when `allowUnsafeSymbols: true` and `strict: true`'
 		);
+		equal(
+			he.encode('äöüÄÖÜ', { 'decimalOutput': true}),
```
Comment (mathiasbynens) [inline tests/tests.src.js]: Please use escapes sequences here, matching the other tests: [`'\xE4\xF6\xFC\xC4\xD6\xDC'`](https://mothereff.in/js-escapes#1%C3%A4%C3%B6%C3%BC%C3%84%C3%96%C3%9C)

File: tests/tests.src.js [hunk 6]
```diff
@@ -6620,6 +6620,11 @@
...
+			'&#228;&#246;&#252;&#196;&#214;&#220;',
+			'encode to decimal HTML entities'
+		);
```
Comment (mathiasbynens) [inline tests/tests.src.js]: Let’s add some more tests that combine `decimal: true` with other options, such as `useNamedReferences`.

**Comment by mathiasbynens** [comment]
Thanks for the patch.

The new setting should also be documented (`README.md`). Once that + the comments I left are addressed this can be merged.

File: tests/tests.src.js [hunk 7]
```diff
@@ -6620,6 +6620,70 @@
...
+			'Does not encode invalid code points whose character references would refer to another code point, neither if `decimal`: true is used'
+		);
+		equal(
+			he.encode('\0\x89', { 'decimal': true, 'encodeEverything': true }),
+			'\0\x89',
+			'Does not encode invalid code points whose character references would refer to another code point, even when `encodeEverything: true` and `decimal: true` is used'
+		);
+		equal(
+			he.encode('foo\xA9<bar\uD834\uDF06>baz\u2603"qux', { 'allowUnsafeSymbols': true }),
+			'foo&#169;<bar&#119558;>baz&#9731;"qux',
```
Comment (akobler) [inline tests/tests.src.js]: This test fails on my local machine. I could not find the reason, since grunt does not print all test output (terminates with: ```Warning: stdout maxBuffer exceeded. Used --force, continuing.```). 

1) Do others have an idea why this test fails? I only replaced the hexadecimal by the decimal representation.
2) How to get all test outputs? Workaround for ```the stdout maxBuffer exceeded``` known? I tried already ```--force``` but without luck.

Comment (mathiasbynens) [inline tests/tests.src.js]: Try in a separate file containing only that test.

**Comment by akobler** [comment]
Thanks for reviewing! All your comments are applied. I fixed the above mentioned failing unit test last week (was really a bug, my changes were not consistent). From my point of view, everything should be OK.
Let me know if something is still missing or if the changes can be merged back.

**Comment by mathiasbynens** [comment]
Thanks. Could you squash your commits please?

File: README.md [hunk 8]
```diff
@@ -104,6 +104,36 @@ he.encode('foo © bar ≠ baz 𝌆 qux', {
...
+he.encode('foo © bar ≠ baz 𝌆 qux', {
+  'decimal': false
+});
+// → 'foo &#xA9; bar &#x2260; baz &#x1D306; qux'
+
+// Passing an `options` object to `encode`, to explicitly enable decimal escapes:
+he.encode('foo © bar ≠ baz 𝌆 qux', {
+  'decimal': true
+});
+// → 'foo &#169; bar &8800; baz &#119558; qux'
```
Comment (mathiasbynens) [inline README.md]: missing `#`

File: man/he.1 [hunk 9]
```diff
@@ -35,6 +35,8 @@ Enable the use of named character references (like `&copy;`) in the output. If c
 Encode every symbol in the input string, even safe printable ASCII symbols.
 .It Sy "--encode --allow-unsafe"
 Encode non-ASCII characters only. This leaves unsafe HTML/XML symbols like `&`, `<`, `>`, `"`, and `'` intact.
+.It Sy "--encode --decimal-output"
+Take decimal format rather than hexadecimal format for encodedHTML entities (`©` is turned into `&#169;` instead of `&#xA9;`)
```
Comment (mathiasbynens) [inline man/he.1]: missing space

**Comment by akobler** [comment]
What's wrong with the last commit? I followed the instructions here: https://github.com/ginatrapani/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit

Did I miss something?

File: src/he.js [hunk 10]
```diff
@@ -150,6 +156,9 @@
 		var encodeEverything = options.encodeEverything;
 		var useNamedReferences = options.useNamedReferences;
 		var allowUnsafeSymbols = options.allowUnsafeSymbols;
+		var escapeCharCode = options.decimal ? decEscape : hexEscape;
+		var escapeSymbol = function(symbol) {return escapeCharCode(symbol.charCodeAt(0));}
```
Comment (mathiasbynens) [inline src/he.js]: Please follow the existing coding style.

File: src/he.js [hunk 11]
```diff
@@ -131,8 +131,14 @@
 		return output;
 	};
 
-	var hexEscape = function(symbol) {
-		return '&#x' + symbol.charCodeAt(0).toString(16).toUpperCase() + ';';
+	// takes integer representing the character code of a symbol
```
Comment (mathiasbynens) [inline src/he.js]: Remove this comment

File: src/he.js [hunk 12]
```diff
@@ -131,8 +131,14 @@
...
+	var hexEscape = function(charCode) {
+		return '&#x' + charCode.toString(16).toUpperCase() + ';';
+	};
+
+	// takes integer representing the character code of a symbol
```
Comment (mathiasbynens) [inline src/he.js]: Remove this comment

File: src/he.js (hunk 10 above)
Comment (mathiasbynens) [inline src/he.js]: Also, let’s declare this outside of the `parseError` block.

**Comment by akobler** [comment]
just fixed your comments
- alignement to code style guide
- comments removed

File: tests/tests.src.js [hunk 13]
```diff
@@ -6620,6 +6620,70 @@
...
+			Error,
+			'Parse error: forbidden code point when `decimal: true`, `strict: true`'
+		);
+		raises(
+			function() {
+				he.encode(<%= stringInvalidCodePoints %>, { 'decimal': true, 'allowUnsafeSymbols': true, 'strict': true });
+			},
+			Error,
+			'Parse error: forbidden code point when `decimal: true`, `allowUnsafeSymbols: true` and `strict: true`'
+		);
```
Comment (mathiasbynens) [inline tests/tests.src.js]: Can you add some tests with astral symbols? E.g. `he.encode('\uD834\uDF06', { 'decimal': true })`.

File: src/he.js [hunk 14]
```diff
@@ -150,6 +154,12 @@
 		var encodeEverything = options.encodeEverything;
 		var useNamedReferences = options.useNamedReferences;
 		var allowUnsafeSymbols = options.allowUnsafeSymbols;
+		var escapeCharCode = options.decimal ? decEscape : hexEscape;
+        
+		var escapeSymbol = function(symbol) {
+            return escapeCharCode(symbol.charCodeAt(0));
+        }
```
Comment (mathiasbynens) [inline src/he.js]: weird indentation, missing semicolon

File: src/he.js [hunk 15]
```diff
@@ -131,8 +131,12 @@
 		return output;
 	};
 
-	var hexEscape = function(symbol) {
-		return '&#x' + symbol.charCodeAt(0).toString(16).toUpperCase() + ';';
+	var hexEscape = function(charCode) {
```
Comment (mathiasbynens) [inline src/he.js]: Let’s use `codePoint` instead of `charCode`.

File: src/he.js [hunk 16]
```diff
@@ -131,8 +131,12 @@
...
+		return '&#x' + charCode.toString(16).toUpperCase() + ';';
+	};
+
+	var decEscape = function(charCode) {
+		return '&#' + charCode + ';';
```
Comment (mathiasbynens) [inline src/he.js]: Same here.

File: man/he.1 [hunk 17]
```diff
@@ -35,6 +35,8 @@ Enable the use of named character references (like `&copy;`) in the output. If c
 Encode every symbol in the input string, even safe printable ASCII symbols.
 .It Sy "--encode --allow-unsafe"
 Encode non-ASCII characters only. This leaves unsafe HTML/XML symbols like `&`, `<`, `>`, `"`, and `'` intact.
+.It Sy "--encode --decimal-output"
+Take decimal format rather than hexadecimal format for encoded HTML entities (`©` is turned into `&#169;` instead of `&#xA9;`).
```
Comment (mathiasbynens) [inline man/he.1]: Let’s reword it as follows:

```
Use the decimal format rather than the hexadecimal format for encoded HTML entities, e.g. `©` is turned into `&#169;` instead of `&#xA9;`.
```

File: tests/tests.src.js [hunk 18]
```diff
@@ -6620,6 +6620,70 @@
 			Error,
 			'Parse error: forbidden code point when `allowUnsafeSymbols: true` and `strict: true`'
 		);
+		equal(
+			he.encode('\xE4\xF6\xFC\xC4\xD6\xDC', { 'decimal': true}),
```
Comment (mathiasbynens) [inline tests/tests.src.js]: missing space before `}` (below too)

File: bin/he [hunk 19]
```diff
@@ -64,6 +64,11 @@
 				options.allowUnsafeSymbols = true;
 				return;
 			}
+			if (string == '--decimal') {
+				action = 'encode';
+				options.decimalOutput = true;
```
Comment (mathiasbynens) [inline bin/he]: Should be `decimal`

File: he.js [hunk 20]
```diff
@@ -142,6 +148,9 @@
 		var encodeEverything = options.encodeEverything;
 		var useNamedReferences = options.useNamedReferences;
 		var allowUnsafeSymbols = options.allowUnsafeSymbols;
+		var escapeCharCode = options.decimal ? decEscape : hexEscape;
```
Comment (mathiasbynens) [inline he.js]: `escapeCodePoint` is more accurate.

File: he.js [hunk 21]
```diff
@@ -142,6 +148,9 @@
...
+		var escapeSymbol = function(symbol) {return escapeCharCode(symbol.charCodeAt(0));}
```
Comment (mathiasbynens) [inline he.js]: `escapeBmpSymbol` is more accurate.

File: man/he.1 [hunk 22]
```diff
@@ -35,6 +35,8 @@ Enable the use of named character references (like `&copy;`) in the output. If c
 Encode every symbol in the input string, even safe printable ASCII symbols.
 .It Sy "--encode --allow-unsafe"
 Encode non-ASCII characters only. This leaves unsafe HTML/XML symbols like `&`, `<`, `>`, `"`, and `'` intact.
+.It Sy "--encode --decimal-output"
```
Comment (mathiasbynens) [inline man/he.1]: `--decimal`

**Comment by mathiasbynens** [comment]
Reviewed + amended your patch and merged it. Thanks!
